bugbug/bug_features.py

```diff
@@ -690,3 +691,185 @@ def apply_rollback(bugs_iter):
+
+        return False
+
+
+class IsPowerBug(SingleBugFeature):
+    """Determine if the bug is related to power based on given bug data."""
+
+    name = "Is Power Bug"
+
+    def __init__(self):
+        self.keywords = set(["power"])
+
+    def __call__(self, bug: bugzilla.BugDict) -> bool:
+        if "[power" in bug["whiteboard"].lower():
+            return True
+
+        if any(
+            keyword.startswith(keyword_start)
+            for keyword_start in self.keywords
+            for keyword in bug["keywords"]
+        ):
+            return True
+
+        return False
+
+
+class IsSecurityBug(SingleBugFeature):
+    """Determine if the bug is related to security based on given bug data."""
+
+    name = "Is Security Bug"
+
+    def __init__(self):
+        self.keywords = set(["sec-", "csectype-"])
+
+    def __call__(self, bug: bugzilla.BugDict) -> bool:
+        if any(
+            f"[{whiteboard_text}" in bug["whiteboard"].lower()
+            for whiteboard_text in ("client-bounty-form", "sec-survey")
+        ):
+            return True
+
+        if any(
+            keyword.startswith(keyword_start)
+            for keyword_start in self.keywords
+            for keyword in bug["keywords"]
+        ):
+            return True
+
+        return False
+
+
+class IsCrashBug(SingleBugFeature):
+    """Determine if the bug is related to crash based on given bug data."""
+
+    name = "Is Crash Bug"
+
+    def __init__(self):
+        self.keywords = set(["crash", "crashreportid"])
+
+    def __call__(self, bug: bugzilla.BugDict) -> bool:
+        if "cf_crash_signature" in bug and bug["cf_crash_signature"] not in ("", "---"):
+            return True
+
+        if any(
+            keyword.startswith(keyword_start)
+            for keyword_start in self.keywords
+            for keyword in bug["keywords"]
+        ):
+            return True
+
+        return False
+
+
+def infer_bug_types(
+    bug: bugzilla.BugDict, bug_map: Optional[dict[int, bugzilla.BugDict]] = None
+) -> list[str]:
+    """Infer bug types based on various bug characteristics.
+
+    Args:
+    - bug (bugzilla.BugDict): A dictionary containing bug data.
+    - bug_map (Optional[dict[int, bugzilla.BugDict]]): A mapping
+        of bug IDs to bug dictionaries. Default is None.
+
+    Returns:
+    - list[str]: A list of inferred bug types (e.g., "memory", "power",
+        "performance", "security", "crash").
+    """
+    is_performance_bug = IsPerformanceBug()
+    is_memory_bug = IsMemoryBug()
+    is_power_bug = IsPowerBug()
+    is_security_bug = IsSecurityBug()
+    is_crash_bug = IsCrashBug()
+
+    types = set()
+
+    if is_memory_bug(bug, bug_map):
+        types.add("memory")
+
+    if is_power_bug(bug):
+        types.add("power")
+
+    if is_performance_bug(bug):
+        types.add("performance")
+
+    if is_security_bug(bug):
+        types.add("security")
+
+    if is_crash_bug(bug):
+        types.add("crash")
+
+    return list(types)
```

Comment:
I would store the functions in a list and loop over it to extract the types.
```suggestion
    return [
        is_type.type_name
        for is_type in self.bug_type_extractors
        if is_type(bug_data)
    ]
```